diff --git a/routers/web/auth/auth.go b/routers/web/auth/auth.go
index 216b8b39be..361500ea92 100644
--- a/routers/web/auth/auth.go
+++ b/routers/web/auth/auth.go
@@ -600,6 +600,11 @@ func handleUserCreated(ctx *context.Context, u *user_model.User, gothUser *goth.
 	notify_service.NewUserSignUp(ctx, u)
 	// update external user information
 	if gothUser != nil {
+		// Inactive OAuth users do not reach handleOAuth2SignIn until after activation.
+		if !u.IsActive {
+			updateAvatarIfNeed(ctx, gothUser.AvatarURL, u)
+		}
+
 		if err := externalaccount.UpdateExternalUser(ctx, u, *gothUser); err != nil {
 			if !errors.Is(err, util.ErrNotExist) {
 				log.Error("UpdateExternalUser failed: %v", err)
